### DISCUSSION TITLE
How to deal with nested C++ objects returned by functions?

### DISCUSSION THREAD

**Discussion by inhahe** [pybind/pybind11#5260] [Q&A]
For example, in the JUCE audio library, AudioPluginInstance has a member getBus() which returns an AudioProcessor::Bus object, which has a member getCurrentLayout(), which returns an AudioChannelSet object, which has a member getChannelTypes(), which returns an array of enums...

So, how would I export that in my dll so that I could do something in Python like, for example.. `p.getBus().getCurrentLayout().getChannelTypes()[0]`?

And of course, on each level of nesting there are also other functions I'd want to access, not just one.

And also, some functions return juce::String objects which pybind11 can't cast automatically, it has to know to call .toStdString on them.

Thanks..

**Comment by jiwaszki** [comment]
@inhahe tl;dr ideally you need to bind everything.

You need to bind **every class that is not stl** in the library, so all of them are properly recognised on some function's return. Refer to this doc for details: https://pybind11.readthedocs.io/en/stable/advanced/cast/index.html

So there are two tricks that you can actually use in case of `juce::String`:
* First one is quite complicated using `type_caster`: https://pybind11.readthedocs.io/en/stable/advanced/cast/custom.html
* Second one is simply wrap around such functions like this:
```
some_class.def("some_func_return_str", [...](SomeClass& cls, ...) {
    return cls.some_func_return_str().toStdString();   // or even wrap it around with py::str(), depends on needs
}
```

If you do not want to bother as much with everything, here some info on automatic binding (it really depends on your needs but from my experience, it is always better to have a fine tuned bindings written yourself): https://pybind11.readthedocs.io/en/stable/compiling.html#generating-binding-code-automatically

**Reply by inhahe** [reply]
Thanks. Before I get too deep into the docs to see how exactly to do things (like, for example, I don't yet know how to bind an object that's only returned as a pointer to it by a C++ function, as opposed to some already-existing static hierarchy), I'm going to check out AutoWIG, because JUCE seems to be a pretty massive library. (After running AutoWIG I'd try editing the generated code to account for String.toStdString() and whatever else I need to change.) But I see that AutoWIG is for Linux, and I'm developing on Windows. I can run it in WSL or something, but I'm not sure if the code it creates will be able to compile and run on Windows/VS 2022. Do you happen to have any idea? Thanks.

**Reply by jiwaszki** [reply]
Tbh I have never used auto-generating tools to such extent. My best bet would be to use something more recent as AutoWIG seems to be not developed for a long time. [robotpy-build](https://github.com/robotpy/robotpy-build) looks like a good place to start, it's only Python dependant so should work anywhere.
